### PR TITLE
Implement Duration.compareTo()

### PR DESCRIPTION
**Pull request by songy23**
Add compareTo() method for Duration. Duration will be ordered similar to Timestamp.

### CODE REVIEW

**Comment by sebright** [comment]
> I didn't implement Comparable or compareTo(), since there could be positive and negative Durations, and I think it's hard to determine the reasonable behavior when we compare a positive Duration with a negative one.

Shouldn't `compareTo` treat positive durations as greater than negative durations?

**Comment by songy23** [comment]
That's my initial thought, but I'm not sure about the use case when users call compareTo(). For example, what if they just want to compare or sort durations by length?

File: api/src/main/java/io/opencensus/common/Duration.java [hunk 1]
```diff
@@ -88,5 +88,21 @@ public static Duration fromMillis(long millis) {
...
+  /**
+   * Compares the length of this {@code Duration} to the specified {@code Duration}.
+   *
+   * @param otherDuration the other {@code Duration} to compare to, not {@code null}.
+   * @return the comparator value: zero if equal, negative if this duration is shorter
+   *     than otherDuration, positive if larger.
+   * @throws NullPointerException if otherDuration is {@code null}.
+   */
+  public int compareLength(Duration otherDuration) {
+    int cmp = compareLong(Math.abs(getSeconds()), Math.abs(otherDuration.getSeconds()));
```
Comment (sebright) [inline api/src/main/java/io/opencensus/common/Duration.java]: Suggestion: You could call `com.google.common.primitives.Longs.compare`.

Comment (sebright) [inline api/src/main/java/io/opencensus/common/Duration.java]: I think that a positive length should be greater than a negative length, similar to a unit-less comparison.

File: core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ void record(List<TagValue> tagValues, double value) {
    */
   double getFraction(Timestamp now) {
     Duration elapsedTime = now.subtractTimestamp(start);
-    checkArgument(toMillis(elapsedTime) >= 0 && toMillis(elapsedTime) < toMillis(duration),
+    checkArgument(toMillis(elapsedTime) >= 0 && elapsedTime.compareLength(duration) < 0,
```
Comment (sebright) [inline core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java]: `toMillis(elapsedTime) >= 0` could compare to a zero-length Duration.

File: api/src/main/java/io/opencensus/common/Duration.java (hunk 1 above)
Comment (songy23) [reply api/src/main/java/io/opencensus/common/Duration.java]: From my understanding lengths are all positive and should be the absolute value of Duration, while Duration could have negative values. I feel it's a little counterintuitive to have negative length.

Comment (songy23) [reply api/src/main/java/io/opencensus/common/Duration.java]: Updated.

File: core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java (hunk 2 above)
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java]: Done.

**Comment by songy23** [comment]
Updated to implement compareTo(), based on our discussion.

**Comment by songy23** [comment]
Thanks!

**Pull request merged**